spec/upload_to_s3_spec.rb

```diff
@@ -160,122 +160,156 @@ def stub_s3_response_for_file(key, exists: true)
-        allow(FastlaneCore::UI).to receive(:important) { |message| warnings << message }
+          warnings = []
+          allow(FastlaneCore::UI).to receive(:important) { |message| warnings << message }
 
-        with_tmp_file(named: 'existing-key') do |file_path|
-          expect(client).to receive(:put_object).with(body: file_instance_of(file_path), bucket: test_bucket, key: expected_key)
+          with_tmp_file(named: 'existing-key') do |file_path|
+            expect(client).to receive(:put_object).with(body: file_instance_of(file_path), bucket: test_bucket, key: expected_key)
 
-          return_value = run_described_fastlane_action(
-            bucket: test_bucket,
-            key: 'existing-key',
-            file: file_path,
-            if_exists: :replace
-          )
+            return_value = run_described_fastlane_action(
+              bucket: test_bucket,
+              key: 'existing-key',
+              file: file_path,
+              if_exists: :replace
+            )
 
-          expect(return_value).to eq(expected_key)
-          expect(Fastlane::Actions.lane_context[Fastlane::Actions::SharedValues::S3_UPLOADED_FILE_PATH]).to eq(expected_key)
+            expect(return_value).to eq(expected_key)
+            expect(warnings).to include("File already exists in S3 bucket #{test_bucket} at #{expected_key}. Will replace with the given one.")
+            expect(Fastlane::Actions.lane_context[Fastlane::Actions::SharedValues::S3_UPLOADED_FILE_PATH]).to eq(expected_key)
+          end
         end
-      end
 
-      it 'throws when if_exists is not one of the expected values' do
-        with_tmp_file(named: 'key') do |file_path|
-          expect do
-            run_described_fastlane_action(
-              bucket: test_bucket,
-              key: 'a8c-key1',
-              file: file_path,
-              if_exists: :invalid
+        it 'throws when if_exists is not one of the expected values' do
+          with_tmp_file(named: 'key') do |file_path|
+            expect do
+              run_described_fastlane_action(
+                bucket: test_bucket,
+                key: 'a8c-key1',
+                file: file_path,
+                if_exists: :invalid
+              )
+            end.to raise_error(FastlaneCore::Interface::FastlaneError, '`if_exist` must be one of :skip, :replace, :fail')
+          end
+        end
+
+        # Untill we remove the deprecated skip_if_exists option, if_exists needs to be optional.
+        # if_exists also has a verify_block that will throw if the given value doesn't match the allowed ones.
+        # This test makes sure that if the user omits or set if_exists as nil, Fastlane bypasses the verify_block.
+        # This is models a possible usage for the action—we want to be extra careful.
+        #
+        # See also https://github.com/wordpress-mobile/release-toolkit/pull/500#discussion_r1239644683
+        it 'does not fail the verify_block if if_exists is explicitly nil' do
```

Comment:
TIL _fastlane_ doesn't call `verify_block` if `nil` is passed. Good to know (and nice to have a test here for confirming this) 👍